samza-core/src/main/java/org/apache/samza/zk/ZkJobCoordinator.java

```diff
@@ -0,0 +1,221 @@
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.samza.zk;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import org.apache.samza.SamzaException;
+import org.apache.samza.config.Config;
+import org.apache.samza.config.JavaSystemConfig;
+import org.apache.samza.coordinator.JobCoordinator;
+import org.apache.samza.coordinator.JobModelManager;
+import org.apache.samza.coordinator.JobModelManager$;
+import org.apache.samza.job.model.JobModel;
+import org.apache.samza.processor.SamzaContainerController;
+import org.apache.samza.system.StreamMetadataCache;
+import org.apache.samza.system.SystemAdmin;
+import org.apache.samza.system.SystemFactory;
+import org.apache.samza.util.SystemClock;
+import org.apache.samza.util.Util;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * JobCoordinator for stand alone processor managed via Zookeeper.
+ */
+public class ZkJobCoordinator implements JobCoordinator, ZkControllerListener {
+  private static final Logger log = LoggerFactory.getLogger(ZkJobCoordinator.class);
+
+  private final ZkUtils zkUtils;
+  private final int processorId;
+  private final ZkController zkController;
+  private final SamzaContainerController containerController;
+  private final BarrierForVersionUpgrade barrier;
+  private final ScheduleAfterDebounceTime debounceTimer;
+  private final StreamMetadataCache  streamMetadataCache;
+  private final ZkKeyBuilder keyBuilder;
+  private final Config config;
+
+  private JobModel newJobModel;
+  private String newJobModelVersion;  // version published in ZK (by the leader)
+  private JobModelManager jobModelManager;
+
+  public ZkJobCoordinator(int processorId, Config config, ScheduleAfterDebounceTime debounceTimer, ZkUtils zkUtils, SamzaContainerController containerController) {
+    this.zkUtils = zkUtils;
+    this.keyBuilder = zkUtils.getKeyBuilder();
+    this.debounceTimer = debounceTimer;
+    this.processorId = processorId;
+    this.containerController = containerController;
+    this.zkController = new ZkControllerImpl(String.valueOf(processorId), zkUtils, debounceTimer, this);
+    this.config = config;
+
+
+    barrier = new ZkBarrierForVersionUpgrade(zkUtils, debounceTimer);
+    streamMetadataCache = initJobModelGeneration();
+  }
+
+  private StreamMetadataCache initJobModelGeneration() {
```

Comment:
method name is misleading as it says "initJobModelGeneration" and all it does is to populate the StreamMetadataCache.